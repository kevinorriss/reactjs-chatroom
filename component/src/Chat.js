import React from 'react'
import autoscroll from 'autoscroll-react'
import Message from './Message'

class Chat extends React.Component {
    render() {
        const { messages, username, ...props } = this.props
        return (
            <div {...props}>
                {messages.map((m, index) => 
                    <Message key={index}
                        isNotification={m.notification}
                        isMe={m.username===username}
                        username={m.username}
                        text={m.text} />
                )}
            </div>
        )
    }
}

export default autoscroll(Chat)
