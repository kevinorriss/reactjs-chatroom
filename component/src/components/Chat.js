import React, { createRef } from 'react'
import Message from './Message'

class Chat extends React.Component {

    constructor(props) {
        super(props)
        this.chatRef = createRef()

        this.autoscroll = true

        this.handleScroll = this.handleScroll.bind(this)
    }

    componentDidUpdate(prevProps) {
        if (prevProps.messages !== this.props.messages && this.autoscroll) {
            this.chatRef.current.scrollTop = this.chatRef.current.scrollHeight - this.chatRef.current.clientHeight
        }
    }

    handleScroll() {
        this.autoscroll = Math.abs(this.chatRef.current.scrollHeight - this.chatRef.current.scrollTop - this.chatRef.current.clientHeight) <= 1
    }

    render() {
        const { messages, username, ...props } = this.props
        return (
            <div {...props} ref={this.chatRef} onScroll={this.handleScroll}>
                {messages.map((m, index) => 
                    <Message key={index}
                        isNotification={m.notification || false}
                        isError={m.error || false}
                        isMe={m.username===username}
                        username={m.username}
                        text={m.text} />
                )}
            </div>
        )
    }
}

export default Chat