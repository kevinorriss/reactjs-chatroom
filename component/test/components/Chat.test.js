// mock autoscroll that simply returns the Chat component
jest.mock('autoscroll-react', () => { return (component) => { return component } })

import Chat from '../../src/components/Chat'

// disable console errors for cleaner test output
jest.spyOn(console, 'error').mockImplementation(() => {})

const username = 'Myself'

const welcome_notification = {
    text: 'Welcome, Myself!',
    notification: true
}

const my_message = {
    text: 'Hello, World!',
    username
}

const received_message = {
    text: 'Hi there!',
    username: 'Other User'
}

const error_message = {
    text: 'There has been an error!',
    error: true
}

const error_and_notification_message = {
    text: 'This is an error and notification...',
    error: true,
    notification: true
}

test('Should render empty chat', () => {
    expect(shallow( <Chat className="chatroom__chat" messages={ [] } username={ username }/> )).toMatchSnapshot()
})

test('Should render notification', () => {
    expect(shallow( <Chat className="chatroom__chat" messages={ [ welcome_notification ] } username={ username }/> )).toMatchSnapshot()
})

test('Should render sent message correctly', () => {
    expect(shallow( <Chat className="chatroom__chat" messages={ [ my_message ] } username={ username }/> )).toMatchSnapshot()
})

test('Should render received message correctly', () => {
    expect(shallow( <Chat className="chatroom__chat" messages={ [ received_message ] } username={ username }/> )).toMatchSnapshot()
})

test('Should render error message correctly', () => {
    expect(shallow( <Chat className="chatroom__chat" messages={ [ error_message ] } username={ username }/> )).toMatchSnapshot()
})

test('Should render error when message is both notification and error', () => {
    expect(shallow( <Chat className="chatroom__chat" messages={ [ error_and_notification_message ] } username={ username }/> )).toMatchSnapshot()
})

test('Should render multiple messages in order', () => {
    expect(shallow( <Chat className="chatroom__chat" messages={ [ welcome_notification, my_message, received_message, error_message ] } username={ username }/> )).toMatchSnapshot()
})