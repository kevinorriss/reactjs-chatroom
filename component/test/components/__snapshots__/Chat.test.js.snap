// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render empty chat 1`] = `
<div
  className="chatroom__chat"
/>
`;

exports[`Should render error message correctly 1`] = `
<div
  className="chatroom__chat"
>
  <Message
    isError={true}
    isMe={false}
    isNotification={false}
    key="0"
    text="There has been an error!"
  />
</div>
`;

exports[`Should render error when message is both notification and error 1`] = `
<div
  className="chatroom__chat"
>
  <Message
    isError={true}
    isMe={false}
    isNotification={true}
    key="0"
    text="This is an error and notification..."
  />
</div>
`;

exports[`Should render multiple messages in order 1`] = `
<div
  className="chatroom__chat"
>
  <Message
    isError={false}
    isMe={false}
    isNotification={true}
    key="0"
    text="Welcome, Myself!"
  />
  <Message
    isError={false}
    isMe={true}
    isNotification={false}
    key="1"
    text="Hello, World!"
    username="Myself"
  />
  <Message
    isError={false}
    isMe={false}
    isNotification={false}
    key="2"
    text="Hi there!"
    username="Other User"
  />
  <Message
    isError={true}
    isMe={false}
    isNotification={false}
    key="3"
    text="There has been an error!"
  />
</div>
`;

exports[`Should render notification 1`] = `
<div
  className="chatroom__chat"
>
  <Message
    isError={false}
    isMe={false}
    isNotification={true}
    key="0"
    text="Welcome, Myself!"
  />
</div>
`;

exports[`Should render received message correctly 1`] = `
<div
  className="chatroom__chat"
>
  <Message
    isError={false}
    isMe={false}
    isNotification={false}
    key="0"
    text="Hi there!"
    username="Other User"
  />
</div>
`;

exports[`Should render sent message correctly 1`] = `
<div
  className="chatroom__chat"
>
  <Message
    isError={false}
    isMe={true}
    isNotification={false}
    key="0"
    text="Hello, World!"
    username="Myself"
  />
</div>
`;
